category/execution/monad/staking/fuzzer/staking_contract_machine.hpp

```diff
@@ -0,0 +1,244 @@
+// Copyright (C) 2025 Category Labs, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include <category/execution/monad/staking/fuzzer/staking_contract_model.hpp>
+#include <category/vm/fuzzing/generator/choice.hpp>
+
+#include <random>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace monad::staking::test
+{
+    class StakingContractMachine
+    {
+        using ValIdSet = vm::fuzzing::UniformSamplingSet<uint64_t>;
+
+        using DelegatorSet = vm::fuzzing::UniformSamplingSet<
+            std::pair<uint64_t, Address>,
+            boost::hash<std::tuple<uint64_t, Address>>>;
+
+        using WithdrawalRequestSet = vm::fuzzing::UniformSamplingSet<
+            std::tuple<uint64_t, Address, uint8_t>,
+            boost::hash<std::tuple<uint64_t, Address, uint8_t>>>;
+
+        using AvailableWithdrawalIdsMap = std::unordered_map<
+            std::tuple<uint64_t, Address>, std::vector<u8_be>,
+            boost::hash<std::tuple<uint64_t, Address>>>;
+
+        StakingContractModel model_;
+        std::mt19937_64 engine_;
+        // Whether the relatively slow public key assertions are enabled:
+        bool enable_pubkey_assertions_;
+        // Whether debug tracing is enabled:
+        bool enable_trace_;
+
+        std::vector<Address> all_addresses_;
```

Comment:
I think these should be tracked in `StakingContractModel` instead as they are properties of the staking contract. Then `StakingContractMachine::model_precompile_*` methods would move into the respective `StakingContractModel` methods.